scripts/solve_network.py

```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
 
-    link_p = get_var(n, "Link", "p")
+    link_p = n.model["Link-p"]
 
     if not electric_ext.empty:
-        link_p_nom = get_var(n, "Link", "p_nom")
+        link_p_nom = n.model["Link-p_nom"]
 
         # ratio of output heat to electricity set by p_nom_ratio
-        lhs = linexpr(
-            (
-                n.links.loc[electric_ext, "efficiency"]
-                * n.links.loc[electric_ext, "p_nom_ratio"],
-                link_p_nom[electric_ext],
-            ),
-            (-n.links.loc[heat_ext, "efficiency"].values, link_p_nom[heat_ext].values),
-        )
+        lhs = (
+            n.links.loc[electric_ext, "efficiency"]
+            * n.links.loc[electric_ext, "p_nom_ratio"]
+            * link_p_nom.loc[electric_ext]
+        ) + (-n.links.loc[heat_ext, "efficiency"] * link_p_nom.loc[heat_ext])
 
-        define_constraints(n, lhs, "=", 0, "chplink", "fix_p_nom_ratio")
+        n.model.add_constraints(lhs == 0, name="chplink-fix_p_nom_ratio")
 
         # top_iso_fuel_line for extendable
-        lhs = linexpr(
-            (1, link_p[heat_ext]),
-            (1, link_p[electric_ext].values),
-            (-1, link_p_nom[electric_ext].values),
+        lhs = (
+            (1 * link_p.loc[heat_ext]),
+            (1 * link_p.loc[electric_ext]),
+            (1 * link_p_nom.loc[electric_ext]),
         )
 
-        define_constraints(n, lhs, "<=", 0, "chplink", "top_iso_fuel_line_ext")
+        n.model.add_contraints(lhs <= 0, name="chplink-top_iso_fuel_line_ext")
 
     if not electric_fix.empty:
         # top_iso_fuel_line for fixed
-        lhs = linexpr((1, link_p[heat_fix]), (1, link_p[electric_fix].values))
+        lhs = (1 * link_p[heat_fix]) + (1 * link_p.loc[electric_fix])
 
-        rhs = n.links.loc[electric_fix, "p_nom"].values
+        rhs = n.links.loc[electric_fix, "p_nom"]
 
-        define_constraints(n, lhs, "<=", rhs, "chplink", "top_iso_fuel_line_fix")
+        n.model.add_constraints(lhs <= rhs, name="chplink-top_iso_fuel_line_fix")
 
     if not electric.empty:
         # backpressure
-        lhs = linexpr(
-            (
-                n.links.loc[electric, "c_b"].values * n.links.loc[heat, "efficiency"],
-                link_p[heat],
-            ),
-            (-n.links.loc[electric, "efficiency"].values, link_p[electric].values),
-        )
+        lhs = (
+            n.links.loc[electric, "c_b"].values
+            * n.links.loc[heat, "efficiency"]
+            * link_p.loc[heat]
+        ) + (-n.links.loc[electric, "efficiency"].values * link_p.loc[electric])
 
-        define_constraints(n, lhs, "<=", 0, "chplink", "backpressure")
+        n.model.add_constraints(lhs <= 0, name="chplink-backpressure")
 
 
 def add_co2_sequestration_limit(n, sns):
     co2_stores = n.stores.loc[n.stores.carrier == "co2 stored"].index
 
-    if co2_stores.empty or ("Store", "e") not in n.variables.index:
+    if co2_stores.empty:  # or ("Store", "e") not in n.variables.index:

```

Comment:
Just for better understanding: what has been the reason of commenting-out the second part of the condiiton?

Reply:
The variables method no longer exists in the network, which is why I commented it out. I also noticed something in pypsa-eur implementation.

Reply:
There seems to be no `variable` method in `n`. I think it relates to the new data structures implemented in `pypsa.Network`

I noticed the same similarities [here](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796/commits/cfb251fc20e4b5a392b077f15f551fb282ed5015) and in pypsa-eur.

Reply:
You mean `n.variables` component, right? 

Agree that some adjustment may be needed, but we have to avoid losses in functionality. It would be great to understand the purpose of having `("Store", "e") not in n.variables.index` condition here, check if it's still needed and find a way to re-implement it. May be a good point to discuss with the cross-sectoral team.

Reply:
I don't know the reason or the functionality. But, I checked similar implementations in pypsa-eur, before the linopy integration, and after the integration, and it seems to have been done in the same way.

Reply:
you can access the variable with `n.model.variables`. so the equivalent check would be `"Store-e" not in n.model.variables`